File: frontend/src/types/race.ts
export interface RaceSummary {
  year: number;
  round: number;
  race_name: string;
  circuit_short_name: string;
  country_name: string;
  date: string;
  total_laps: number;
  safety_car_count: number;
  red_flag_count: number;
  winner: {
    driver_number: number;
    full_name: string;
    name_acronym: string;
    team_name: string;
    team_colour: string;
  };
}

export interface DriverRaceResult {
  position: number;
  driver_number: number;
  full_name: string;
  name_acronym: string;
  team_name: string;
  team_colour: string;
  grid: number;
  positions_gained: number;
  points: number;
  status: string;
  total_time: string | null;
  fastest_lap_time: string | null;
  pit_count: number;
  headshot_url: string | null;
}

export interface LapData {
  driver_number: number;
  name_acronym: string;
  team_colour: string;
  lap_number: number;
  lap_duration: number | null;
  sector_1: number | null;
  sector_2: number | null;
  sector_3: number | null;
  compound: TyreCompound;
  tyre_age: number;
  is_pit_out_lap: boolean;
}

export interface PositionData {
  driver_number: number;
  name_acronym: string;
  team_colour: string;
  laps: { lap: number; position: number }[];
}

export interface PitStrategy {
  driver_number: number;
  name_acronym: string;
  team_name: string;
  team_colour: string;
  stints: Stint[];
  pit_stops: PitStop[];
}

export interface Stint {
  stint_number: number;
  compound: TyreCompound;
  lap_start: number;
  lap_end: number;
  stint_laps: number;
  tyre_age_at_start: number;
}

export interface PitStop {
  lap_number: number;
  pit_duration: number;
}

export type TyreCompound =
  | "SOFT"
  | "MEDIUM"
  | "HARD"
  | "INTERMEDIATE"
  | "WET";

export interface RaceListItem {
  session_key: number;
  year: number;
  round: number;
  country_name: string;
  country_code: string;
  circuit_short_name: string;
  location: string;
  date_start: string;
  session_type: string;
  session_name: string;
  winner_name: string | null;
  winner_team: string | null;
  winner_team_colour: string | null;
}

File: backend/app/models/race.py
from pydantic import BaseModel


# --- Race List ---

class RaceListItem(BaseModel):
    session_key: int
    year: int
    round: int
    country_name: str
    country_code: str | None = None
    circuit_short_name: str
    location: str
    date_start: str
    session_type: str
    session_name: str
    winner_name: str | None = None
    winner_team: str | None = None
    winner_team_colour: str | None = None


class SeasonRacesResponse(BaseModel):
    year: int
    races: list[RaceListItem]


# --- Race Summary ---

class WinnerInfo(BaseModel):
    driver_number: int | None = None
    full_name: str
    name_acronym: str
    team_name: str
    team_colour: str


class RaceSummaryResponse(BaseModel):
    year: int
    round: int
    race_name: str
    circuit_short_name: str
    country_name: str
    date: str
    total_laps: int
    safety_car_count: int
    red_flag_count: int
    winner: WinnerInfo


# --- Race Results ---

class DriverRaceResult(BaseModel):
    position: int | None = None
    driver_number: int | None = None
    full_name: str
    name_acronym: str
    team_name: str
    team_colour: str
    grid: int | None = None
    positions_gained: int | None = None
    points: float = 0.0
    status: str | None = None
    total_time: str | None = None
    fastest_lap_time: str | None = None
    pit_count: int = 0
    headshot_url: str | None = None


class RaceResultsResponse(BaseModel):
    results: list[DriverRaceResult]


# --- Lap Data ---

class LapData(BaseModel):
    driver_number: int
    name_acronym: str
    team_colour: str
    lap_number: int
    lap_duration: float | None = None
    sector_1: float | None = None
    sector_2: float | None = None
    sector_3: float | None = None
    compound: str | None = None
    tyre_age: int = 0
    is_pit_out_lap: bool = False


class LapsResponse(BaseModel):
    laps: list[LapData]


# --- Position Data ---

class LapPosition(BaseModel):
    lap: int
    position: int


class PositionData(BaseModel):
    driver_number: int
    name_acronym: str
    team_colour: str
    laps: list[LapPosition]


class PositionsResponse(BaseModel):
    positions: list[PositionData]


# --- Session Detail ---

class SessionDetail(BaseModel):
    session_key: int
    year: int
    session_name: str | None = None
    session_type: str | None = None
    circuit_short_name: str | None = None
    country_name: str | None = None
    country_code: str | None = None
    location: str | None = None
    date_start: str | None = None
    date_end: str | None = None
